jobs: fix semantics of PAUSE JOB

spaskob opened the pull request:
So far PAUSE JOB would mark the job as paused
immediately but without actually pausing it.
Instead it relied on the job trying to write
progress at some point and fail. We introduce
a new job state StatusPauseRequested to fix that.
It is used to designate a job that is still running but
has been requested to be paused.

Additionally:
We also make jobs that are reverting pausable.
A few tests that relied on PAUSE pausing the job
immediately after returning were fixed.

Fixes #30505.
Fixes #45809 

Release note (bug fix): 19.2 `IMPORT INTO` jobs could be
pause while recovering from a failure and subsequently resumed
as if the failure did not occur.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/44558)
<!-- Reviewable:end -->

spaskob committed:
jobs: fix semantics of PAUSE JOB

So far PAUSE JOB would mark the job as paused
immediately but without actually pausing it.
Instead it relied on the job trying to write
progress and some point and fail. We introduce
a new job state StatusPauseRequested to fix that.
It marks a job that is still running but has been
requested to be paused.

We also make jobs that are reverting pausable.
A few tests that relied on PAUSE pausing the job
immediately after returning were fixed.

Release note: none.

rytaft commented:
Is this ready for review? The PR message is still empty





<!-- Sent from Reviewable.io -->

spaskob commented:
Thanks, yes it is. I completely forgot to copy it from the commit.

On Thu, Feb 6, 2020 at 1:27 PM Rebecca Taft <notifications@github.com>
wrote:

> Is this ready for review? The PR message is still empty
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/44558?email_source=notifications&email_token=AANL635FTPOU5BU6SBLNLK3RBRJB3A5CNFSM4KN4ZOQ2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOELAIYWQ#issuecomment-583044186>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANL633F4MJJEU4Q5F4MG3LRBRJB3ANCNFSM4KN4ZOQQ>
> .
>

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but I'm not familiar with most of this code. Would be good to get at least one more set of eyes.

> Reviewed 8 of 8 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44558#-:-M-kOOpL2_PZFo4_QPwM:bwonayq)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @danhhz, @lucy-zhang, and @spaskob)

___
*[pkg/ccl/changefeedccl/cdctest/testfeed.go, line 269 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44558#-M-k3Fib4U2bSRZjpPM2:-M-k3Fib4U2bSRZjpPM3:b-ss7aur) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bb394dfe2ececf01b33c68107047f9e9d8ce3b5/pkg/ccl/changefeedccl/cdctest/testfeed.go#L269)):*
> ```Go
> 	}
> 	// PAUSE JOB does not actually pause the job bt only sends a request
> 	// it. Actually block until the job state changes.
> ```

bt -> but
request it -> request for it
___
*[pkg/ccl/changefeedccl/cdctest/testfeed.go, line 276 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44558#-M-kMrX4--VwOFvzsK5z:-M-kMrX4--VwOFvzsK6-:b-idlspw) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bb394dfe2ececf01b33c68107047f9e9d8ce3b5/pkg/ccl/changefeedccl/cdctest/testfeed.go#L276)):*
> ```Go
> 	}
> 	ctx := context.Background()
> 	return retry.WithMaxAttempts(ctx, opts, 10, func() error {
> ```

Have you seen `testutils.SucceedsSoon`? That may allow to you to achieve what you want here without configuring the `retry.Options` (but I don't think it matters too much).
___
*[pkg/jobs/jobs_test.go, line 448 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44558#-M-kMNDsE9_ma08BRrp4:-M-kMNDsE9_ma08BRrp5:bmx4ed4) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bb394dfe2ececf01b33c68107047f9e9d8ce3b5/pkg/jobs/jobs_test.go#L448)):*
> ```Go
> 			}
> 			// Not committed yet, so state shouldn't have changed.
> 			// Don' check status in txn.
> ```

Don' -> Don't
___
*[pkg/jobs/jobs_test.go, line 485 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44558#-M-kNCvW7YuyU2vbsdA1:-M-kNCvW7YuyU2vbsdA2:bmx4ed4) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bb394dfe2ececf01b33c68107047f9e9d8ce3b5/pkg/jobs/jobs_test.go#L485)):*
> ```Go
> 			}
> 			// Not committed yet, so state shouldn't have changed.
> 			// Don' check status in txn.
> ```

Don' -> Don't
___
*[pkg/sql/stats/create_stats_job_test.go, line 359 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44558#-M-k5-z7-RbHISjYOU0t:-M-k5-z7-RbHISjYOU0u:b65ppn1) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbeb04dae240e93222a87dd6aea89db4cd3735cc/pkg/sql/stats/create_stats_job_test.go#L359)):*
> ```Go
> 	}
> 
> 	// Attempt to start an automatic stats run. It should fail.
> ```

I think this comment should stay
___
*[pkg/sql/stats/create_stats_job_test.go, line 366 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44558#-M-k563o3stb8JcKBzEG:-M-k563pCDziQUyB9Cn0:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbeb04dae240e93222a87dd6aea89db4cd3735cc/pkg/sql/stats/create_stats_job_test.go#L366)):*
> ```Go
> 	}
> 
> 	// Pause the job. Starting another automatic stats run should still fail.
> ```

ditto
___
*[pkg/sql/stats/create_stats_job_test.go, line 677 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44558#-M-k5DVw-q6I44Fp4nne:-M-k5DVw-q6I44Fp4nnf:bf1w8g5) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bb394dfe2ececf01b33c68107047f9e9d8ce3b5/pkg/sql/stats/create_stats_job_test.go#L677)):*
> ```Go
> 				// Read from the channel twice to allow jobutils.RunJob to complete
> 				// even though there is only one ScanRequest. Add a timeout to detect
> 				// deadlock due to faulty test logic.
> ```

Where is the timeout?


<!-- Sent from Reviewable.io -->

spaskob reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44558)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @danhhz, @lucy-zhang, @rytaft, and @spaskob)

___
*[pkg/ccl/changefeedccl/cdctest/testfeed.go, line 269 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44558#-M-k3Fib4U2bSRZjpPM2:-M-kVJjB6p25hh044W5q:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bb394dfe2ececf01b33c68107047f9e9d8ce3b5/pkg/ccl/changefeedccl/cdctest/testfeed.go#L269)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

bt -> but
request it -> request for it
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/cdctest/testfeed.go, line 276 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44558#-M-kMrX4--VwOFvzsK5z:-M-kVK3X9Qa5qG_8do0P:b-ys5nmb) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bb394dfe2ececf01b33c68107047f9e9d8ce3b5/pkg/ccl/changefeedccl/cdctest/testfeed.go#L276)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Have you seen `testutils.SucceedsSoon`? That may allow to you to achieve what you want here without configuring the `retry.Options` (but I don't think it matters too much).
</blockquote></details>

Thanks useful to have in the toolbox. 45s is a too much here and also I am switching PAUSE to be blocking this week so I will just remove this code after that.
___
*[pkg/jobs/jobs_test.go, line 448 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44558#-M-kMNDsE9_ma08BRrp4:-M-kVf1z3dMjBh8czD8o:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bb394dfe2ececf01b33c68107047f9e9d8ce3b5/pkg/jobs/jobs_test.go#L448)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Don' -> Don't
</blockquote></details>

Done.
___
*[pkg/jobs/jobs_test.go, line 485 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44558#-M-kNCvW7YuyU2vbsdA1:-M-kVjwX-b5EPF_8uuUC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bb394dfe2ececf01b33c68107047f9e9d8ce3b5/pkg/jobs/jobs_test.go#L485)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Don' -> Don't
</blockquote></details>

Done.
___
*[pkg/sql/stats/create_stats_job_test.go, line 359 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44558#-M-k5-z7-RbHISjYOU0t:-M-kXmTcCYR-lZallc54:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbeb04dae240e93222a87dd6aea89db4cd3735cc/pkg/sql/stats/create_stats_job_test.go#L359)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I think this comment should stay
</blockquote></details>

Done.
___
*[pkg/sql/stats/create_stats_job_test.go, line 677 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44558#-M-k5DVw-q6I44Fp4nne:-M-kVvWY5iuBRrkTDH_g:b-6exldr) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bb394dfe2ececf01b33c68107047f9e9d8ce3b5/pkg/sql/stats/create_stats_job_test.go#L677)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Where is the timeout?
</blockquote></details>

thanks, I decided no to add it.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
> Reviewed 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44558)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @danhhz and @lucy-zhang)



<!-- Sent from Reviewable.io -->

spaskob commented:
bors r+

spaskob commented:
bors r+
